hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncNonMetaRegionLocator.java

```diff
@@ -217,6 +217,223 @@ private boolean tryComplete(LocateRequest req, CompletableFuture<RegionLocations
+              + "new location {} (endKey={})",
+            overlap.getValue(),
+            Bytes.toStringBinary(overlap.getValue().getRegionLocation().getRegion().getEndKey()),
+            locations, Bytes.toStringBinary(locations.getRegionLocation().getRegion().getEndKey()));
+        }
+
+        cache.remove(overlap.getKey());
+      }
+    }
+
+    private boolean isEqual(RegionLocations locs1, RegionLocations locs2) {
+      HRegionLocation[] locArr1 = locs1.getRegionLocations();
+      HRegionLocation[] locArr2 = locs2.getRegionLocations();
+      if (locArr1.length != locArr2.length) {
+        return false;
+      }
+      for (int i = 0; i < locArr1.length; i++) {
+        // do not need to compare region info
+        HRegionLocation loc1 = locArr1[i];
+        HRegionLocation loc2 = locArr2[i];
+        if (loc1 == null) {
+          if (loc2 != null) {
+            return false;
+          }
+        } else {
+          if (loc2 == null) {
+            return false;
+          }
+          if (loc1.getSeqNum() != loc2.getSeqNum()) {
+            return false;
+          }
+          if (!Objects.equal(loc1.getServerName(), loc2.getServerName())) {
+            return false;
+          }
+        }
+      }
+      return true;
+    }
+
+    /** Returns all cached RegionLocations */
+    public Collection<RegionLocations> getAll() {
+      return cache.values();
+    }
+
+    /**
+     * Gets the RegionLocations for a given region's startKey. This is a direct lookup, if the key
+     * does not exist in the cache it will return null.
+     * @param startKey region start key to directly look up
+     */
+    public RegionLocations get(byte[] startKey) {
+      return cache.get(startKey);
+    }
+
+    /**
+     * Finds the RegionLocations for the region with the greatest startKey less than or equal to the
+     * given row
+     * @param row row to find locations
+     */
+    public RegionLocations findForRow(byte[] row) {
+      Map.Entry<byte[], RegionLocations> entry = cache.floorEntry(row);
+      if (entry == null) {
+        return null;
+      }
+      return entry.getValue();
+    }
+
+    /**
+     * Finds the RegionLocations for the region with the greatest startKey strictly less than the
+     * given row
+     * @param row row to find locations
+     */
+    public RegionLocations findForBeforeRow(byte[] row) {
+      boolean isEmptyStopRow = isEmptyStopRow(row);
+      Map.Entry<byte[], RegionLocations> entry =
+        isEmptyStopRow ? cache.lastEntry() : cache.lowerEntry(row);
+      if (entry == null) {
+        return null;
+      }
+      return entry.getValue();
+    }
+
+    /**
+     * Removes the location from the cache if it exists and can be removed. Once a change is deemed
+     * possible, calls beforeUpdate callback prior to making a change. Calls afterUpdate callback
+     * after making a change.
+     */
+    public synchronized void remove(HRegionLocation loc, Runnable beforeUpdate,
```

Comment:
We need to be careful that the beforeUpdate and afterUpdate do not hold other locks otherwise it may introduce dead lock

Reply:
Let me see if I can remove the callbacks. I was trying to keep the metaLocation.onError stuff out of here. I wasn't sure if the onError call needed to happen at the exact point

Reply:
If it is not easy, just add more comments to warn others.

Reply:
I removed the callbacks. I think `remove` can just return a boolean, and both actions can just happen after if an action was taken.